torchtune/utils/_checkpointing/_utils.py

```diff
@@ -0,0 +1,53 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+# All rights reserved.
+#
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+
+from typing import Dict, Iterable, Tuple
+from warnings import warn
+
+import torch
+
+
+def update_state_dict_for_classifier(
+    state_dict: Dict[str, torch.Tensor],
+    model_named_parameters: Iterable[Tuple[str, torch.nn.Parameter]],
+):
+    """
+    Validates the state dict for checkpoint loading for a classifier model, to be
+    used prior to a call to ``model.load_state_dict(state_dict)``.
+    When fine-tuning a classifier model from the checkpoint of a base language model
+    which has ``output.weight`` of shape ``[vocab_dim, embed_dim]``, we overwrite
```

Comment:
super nit: we dont check for [vocab_dim, embed_dim]`, so it could be of any shape (except the checkpoint's shape). Is that correct?

Just for inspiration, in case you decide to make changes, maybe we could do something like below (this is really just an idea. I am not asking you to replace the docstring :P )

"""
Used to replace the output.weight of a base model when the checkpoint output.weight has a different shape.

This is useful for situations like reward modeling, when the base model originally has shape ``[vocab_dim, embed_dim]``, but we want it to be something like ``[num_classes, embed_dim]``.

Notes:
- output.bias will be ignored
- If the output.shape is already of the same shape as the checkpoint, this function doesn't replace it.
"""

Btw, just confirming: The reason we only swap the weights if their shape is different is because the ``model.load_state_dict`` will have already loaded the weights correctly?

Reply:
> The reason we only swap the weights if their shape is different is because the model.load_state_dict will have already loaded the weights correctly?

This is a good point, maybe we always want to swap the weights. If the user is starting from another classifier model with the same `num_classes` but is using it for a different task, does it make sense to start off with the pretrained weights?

> super nit: we dont check for [vocab_dim, embed_dim]`, so it could be of any shape (except the checkpoint's shape). Is that correct?

Sorry, what do you mean here? We're assuming the base model is identical aside from the output layer.

Reply:
To answer your question:
>  The reason we only swap the weights if their shape is different is because the model.load_state_dict will have already loaded the weights correctly?

this is before load state dict